feat(webui): dark mode as in #179

Shadowfied opened the pull request:
I added a toggle for Vuetifys dark theme as previously mentioned in #179. I looked through all UI elements I could find in the frontend in an empty library and in a library with some items added and fixed up any oddities.

gotson commented:
Didn't have a look at the code yet, but tried the rendering.

I think the primary color (blue) doesn't work that well with the dark theme. We may need to adjust it, or use a lighter tint.
<img width="263" alt="Screenshot 2020-07-22 at 19 25 30" src="https://user-images.githubusercontent.com/2139133/88175458-74fac000-cc58-11ea-889a-9072e775de48.png">


The drop shadow on the cards and lists are not showing very clearly. This is really a material design thing, i have to check if we can do something about it.

<img width="565" alt="Screenshot 2020-07-22 at 19 23 28" src="https://user-images.githubusercontent.com/2139133/88175540-92c82500-cc58-11ea-9420-78f4ff8cc695.png">
<img width="568" alt="Screenshot 2020-07-22 at 19 23 36" src="https://user-images.githubusercontent.com/2139133/88175547-952a7f00-cc58-11ea-9c69-3a9be5597d99.png">


The highlighting on cards when doing a multi-select is not showing, and the selection icon color is grey also:

<img width="162" alt="Screenshot 2020-07-22 at 19 20 46" src="https://user-images.githubusercontent.com/2139133/88171129-665cda80-cc51-11ea-8eb5-326d86296a65.png">
<img width="161" alt="Screenshot 2020-07-22 at 19 20 54" src="https://user-images.githubusercontent.com/2139133/88171141-6957cb00-cc51-11ea-9151-911b89216ef8.png">
<img width="59" alt="Screenshot 2020-07-22 at 20 24 13" src="https://user-images.githubusercontent.com/2139133/88176103-5fd26100-cc59-11ea-8983-8956a4abc963.png">


Regarding the menu item, we could use _Light theme_ / _Dark theme_ for the wording instead of on/off. We may also use a different icon depending if it's light or not.

Shadowfied commented:
Thanks for your feedback!

About the primary color, it passes Chromes contrast check, but it could be better. We can easily get around that by specifying another color for primary in dark:

```
    themes: {
      light: {
        primary: '#005ed3',
        secondary: '#fec000',
        accent: '#ff0335',
      },
      dark: {
        primary: '#78baec',
      },
    },
```

As for the box shadows using elevation, looks like Vuetify doesn't natively support different colors for different themes: https://github.com/vuetifyjs/vuetify/issues/7147
There seem to be workarounds but I don't think the lack of drop shadows will prevent usability for the dark mode.

I also fixed the outlines and the checkmark button etc. I assumed the dark theme would inherit the primary, secondary and accent, but turns out that's not the case, so I added the secondary and accent colors to the dark theme as well, which fixes it:

```
    themes: {
      light: {
        primary: '#005ed3',
        secondary: '#fec000',
        accent: '#ff0335',
      },
      dark: {
        primary: '#78baec',
        secondary: '#fec000',
        accent: '#ff0335',
      },
```

primetoxinz commented on komga-webui/src/components/ItemBrowser.vue:
```diff
@@ -178,7 +178,7 @@ export default Vue.extend({
   methods: {
     onResize () {
       const content = this.$refs.content as HTMLElement
-      this.width = computeCardWidth(content.clientWidth, this.$vuetify.breakpoint.name)
```
What's with all the `as any` casts?

gotson commented:
> Moon for night theme and sun for light theme?

Not a big fan of that, we can stick with _Light theme_ / _Dark theme_.

gotson commented on komga-webui/src/components/bars/ToolbarSticky.vue:
```diff
@@ -16,22 +16,25 @@ export default Vue.extend({
       type: Number,
       default: undefined,
     },
-    color: {
```
Removing this prop will make the component less reusable. Vuetify offers [custom colors](https://vuetifyjs.com/en/customization/theme/#custom-theme-variants), we could define a base color for each theme, based on white/black, which would have lighten/darken flavors. We would need to invert the lighten/darken for the black theme, then we could use that by using `base darken-1` for example, and it would adjust automatically whatever the theme we use.

gotson commented on komga-webui/src/components/bars/ToolbarSticky.vue:
```diff
@@ -16,22 +16,25 @@ export default Vue.extend({
   computed: {
     barStyle (): any {
-      if (['xs', 'sm'].includes(this.$vuetify.breakpoint.name)) {
+      if (['xs', 'sm'].includes(this.$vuetify.breakpoint.name as any)) {
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/components/ItemBrowser.vue:
```diff
@@ -178,7 +178,7 @@ export default Vue.extend({
     onResize () {
       const content = this.$refs.content as HTMLElement
-      this.width = computeCardWidth(content.clientWidth, this.$vuetify.breakpoint.name)
+      this.width = computeCardWidth(content.clientWidth, this.$vuetify.breakpoint.name as any)
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/views/BookReader.vue:
```diff
@@ -627,7 +627,7 @@ export default Vue.extend({
           page: this.currentPage.toString(),
         },
-      })
+      } as any)
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/views/BrowseCollections.vue:
```diff
@@ -163,7 +163,7 @@ export default Vue.extend({
           pageSize: `${this.pageSize}`,
         },
-      }).catch(_ => {
+      } as any).catch(_ => {
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/views/BrowseLibraries.vue:
```diff
@@ -309,7 +309,7 @@ export default Vue.extend({
           readStatus: `${this.filters.readStatus}`,
         },
-      }).catch(_ => {
+      } as any).catch(_ => {
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/views/BrowseSeries.vue:
```diff
@@ -323,7 +323,7 @@ export default Vue.extend({
           readStatus: `${this.filters.readStatus}`,
         },
-      }).catch(_ => {
+      } as any).catch(_ => {
```
cast as `any` shouldn't be necessary

gotson commented on komga-webui/src/views/Home.vue:
```diff
@@ -97,6 +97,20 @@
+            <v-icon>mdi-lightbulb</v-icon>
+          </v-list-item-icon>
+          <v-list-item-content>
+            <v-list-item-title v-if="this.$vuetify.theme.dark">Dark mode off</v-list-item-title>
+            <v-list-item-title v-else>Dark mode on</v-list-item-title>
```
we can change the wording to _Light theme_ / _Dark theme_

gotson commented on komga-webui/src/views/Home.vue:
```diff
@@ -97,6 +97,20 @@
+    <v-list>
+        <v-list-item @click="toggleDarkMode">
+          <v-list-item-icon>
+            <v-icon>mdi-lightbulb</v-icon>
```
The icon could change depending on the theme in use.

Maybe `mdi-brightness-7` when the light theme is on, and `mdi-brightness-3` when the dark theme is on.

gotson commented on komga-webui/src/views/Home.vue:
```diff
@@ -121,19 +135,28 @@ import LibraryActionsMenu from '@/components/menus/LibraryActionsMenu.vue'
     }
+
+    if (this.$cookies.isKey(cookieDarkMode)) {
+      const darkMode = Boolean(JSON.parse(this.$cookies.get(cookieDarkMode)))
```
why do you need to use `JSON.parse` when getting the cookie?

gotson commented on komga-webui/src/views/Home.vue:
```diff
@@ -121,19 +135,28 @@ import LibraryActionsMenu from '@/components/menus/LibraryActionsMenu.vue'
+
+    if (this.$cookies.isKey(cookieDarkMode)) {
+      const darkMode = Boolean(JSON.parse(this.$cookies.get(cookieDarkMode)))
+      this.darkMode = darkMode
+      this.$vuetify.theme.dark = darkMode
```
why do you need `this.darkMode` when you already have the vuetify prop?

gotson commented on komga-webui/src/views/Home.vue:
```diff
@@ -147,6 +170,12 @@ export default Vue.extend({
+    toggleDarkMode () {
+      this.darkMode = !this.darkMode
+
+      this.$cookies.set(cookieDarkMode, JSON.stringify(this.darkMode), Infinity)
```
why do you need to `JSON.stringify` a boolean?

Shadowfied commented:
Committed the fixes for the theme I referred to in my previous post.


Regarding all the "as any" casts, I initially installed the packages and ran the scripts using yarn which I primarily use cause it's faster. I didn't realize it used anything different than npm, but upon installing and serving with yarn, TypeScript errors. [Here](https://pastebin.com/deqQbqfQ) if you're curious.
I removed node_modules and installed using npm and then the errors are gone, so I've reverted all the "as any" casts.


> Removing this prop will make the component less reusable. Vuetify offers [custom colors](https://vuetifyjs.com/en/customization/theme/#custom-theme-variants), we could define a base color for each theme, based on white/black, which would have lighten/darken flavors. We would need to invert the lighten/darken for the black theme, then we could use that by using `base darken-1` for example, and it would adjust automatically whatever the theme we use.

After some discussion on Discord we decided on adding a new contrast color and use that here for now. Done in latest commit.

> Maybe mdi-brightness-7 when the light theme is on, and mdi-brightness-3 when the dark theme is on.

Done in latest commit!

> why do you need to use JSON.parse when getting the cookie?

Because cookies are string valued. If we retrieve a cookie with the string "true" or "false" we can't use or cast that to a boolean. This seems like the simplest and cleanest solution but let me know if there's a better way to go about it and I can do that instead.

> why do you need this.darkMode when you already have the vuetify prop?

Not sure why I did that. Removed in latest commit.

gotson commented:
The multi-select bar is not switching to dark theme:

<img width="631" alt="Screenshot 2020-07-24 at 10 02 35" src="https://user-images.githubusercontent.com/2139133/88354861-ecce0500-cd94-11ea-818a-4176b244064f.png">
<img width="791" alt="Screenshot 2020-07-24 at 10 02 28" src="https://user-images.githubusercontent.com/2139133/88354863-ee97c880-cd94-11ea-8e2b-0fcf814eb65b.png">

Same for the edit bar when editing elements of a collection.

Those bars are based on the `Toolbar-Sticky` and use the `color` prop we mentioned before.

I added a `base` color in our palette:
```js
light: {
        base: colors.shades.white,
      },
      dark: {
        base: colors.shades.black,
      },
```

Then i changed the `color` prop for the `toolbar-sticky` element to `base` in `SeriesMultiSelectBar`, `BooksMultiSelectBar`, and in `BrowseCollection`.

---

I also checked the empty state avatar, with the new `contrast-1` color it looks good on both themes, i think you can put it back.

Shadowfied commented:
Good catch. Is there a point in us using base as color for those bars though? I fully support reusability in colors, but I feel like if we're gonna leave the default to contrast-1 which is what we're using in most cases, since it's essentially the same context for SeriesMultiSelectBar, BooksMultiSelectBar and BrowseCollection, isn't it better to leave off the color prop in all these cases so we're consistent? Either that or we apply base to all our bars, but I think the contrast color looks better given the context.

gotson commented:
You have to distinguish reusable ui components from instantiated components, the sticky bar need the color prop to be customizable.

Adding base in our color palette will help to be theme agnostic that's all.

The multi select bars that appear on top of the existing bar actually have a different color, not the contrast-1 but the base. But they could use something different.

Shadowfied commented:
You misunderstand me! 

I agree that we should have the color prop. I'm just saying, is it worth to actually **use the prop** in the cases of SeriesMultiSelectBar, BooksMultiSelectBar and BrowseCollection, since we're letting, for example BrowseSeries use default. I'm just asking if we should let all the current instances of ToolbarSticky use the default, for consistency. Cause if we do as you proposed with changing the "white" color to "base", we'll have different looking ToolbarSticky in BrowseSeries and BrowseCollection for example.

I just think since they are all in the same position and context aesthetically, they should all be using contrast-1 for now, meaning we should not give them an explicit color at all, but let them use the default.

BrowseLibraries:
<img width="380" alt="Screenshot 2020-07-24 at 14 29 33" src="https://user-images.githubusercontent.com/6698301/88391251-1baaa200-cdba-11ea-9a93-9f9cad8933a9.png">

BrowseLibraries during SeriesMultiSelectBar:
<img width="410" alt="Screenshot 2020-07-24 at 14 30 04" src="https://user-images.githubusercontent.com/6698301/88391288-2cf3ae80-cdba-11ea-8f85-9c2bacdc736e.png">

Or was your intention for them to be differentiated by color? If so, I agree with the decision for SeriesMultiSelectBar and BooksMultiSelectBar as they present a new context, but BrowseCollection should probably keep the default like BrowseLibraries.

gotson commented:
Sorry! 😅

The selection bars were white, on purpose. The different color helps to visualize that its not the same toolbar. 

But the black color may not be right for the dark theme indeed. 

I would like to keep them different in color, but I'm open for the dark theme color instead of black, and we could name the color something other than base also!

Shadowfied commented:
All good :)

In that case, I think the black works fine actually, however, I do think we should let it use the default color in "BrowseCollection", as that is not the action toolbar, it's the same context as the other Browse- screens, and wouldn't visualize the change upon doing a multi select in the view.

I committed my changes according to this. Please see what you think!

gotson commented:
Note that there are 2 bars in BrowseCollection, the default toolbar showing the name, and another bar that appears on top when you click the edit items button. The second one is the one I was talking about, which is currently white, and has the same behavior as the multi select bars for books and series. 

![Screenshot_20200724-220419__01](https://user-images.githubusercontent.com/2139133/88399902-2553fa00-cdfa-11ea-9a7a-6b6f1fd5be38.jpg)
![Screenshot_20200724-220433](https://user-images.githubusercontent.com/2139133/88399908-271dbd80-cdfa-11ea-8f32-a80b4a3dfd81.jpg)

Shadowfied commented:
Ah, missed that one. Adjusted it.

gotson commented:
:tada: This PR is included in version 0.49.0 :tada:

The release is available on:
- `v0.49.0`
- [GitHub release](https://github.com/gotson/komga/releases/tag/v0.49.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
